Add modpack update page in instance settings

flowln opened the pull request:
Closes #180
Closes #170

~TODO:~ I'm satisfied with the state of this for now, so I'll open it for review :p

- [x] Modrinth modpack page
- [x] Flame modpack page
- [x] Fix icons with managed packs not from MR/CF (or probably hide them since we can't update them yet anyways)
- [x] Make the changelog thing parse the content as markdown
- [x] Add link to the modpack page in the mod provider
    - [x] Modrinth URLs
    - [ ] ~Flame URLs~ not really sure how to do this. Maybe we should have another field in `instance.cfg`, or do another NetJob to get the thing (though it would be slow, blocking and really not super useful)

and probably some more stuff i forgot about :x

Status (as of https://github.com/PrismLauncher/PrismLauncher/pull/32/commits/9cf53f847315726688a710d14bc47671f6cac597):

![FO modpack page](https://user-images.githubusercontent.com/9145768/201484408-14d30f1d-5405-4526-96b1-2ea30818ee2f.png)

![Nomifactory modpack page](https://user-images.githubusercontent.com/9145768/201484623-d1e86ca8-fb6d-492f-b1a5-bb5d454f73b8.png)

flowln committed:
feat(ui): add ManagedPackPage ui

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: add basic ManagedPackPage classes

The idea is to have a base class that defines common behavior, and
subclasses for each modpack provider, adding specific behavior.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: add logic for the modrinth instance modpack page

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: add (current) indicator to the currently installed version in MR

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: add PageContainer::getPage

This allows us to directly access a page from outside. This will be
useful for telling the ManagedPackPage who is the window it's on, so
that we can close it when updating :^)

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
refactor: abstract away update confirmation dialog

... so that we can avoid code duplication.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: allow skipping the update confirmation dialog

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: add functionality to MR modpack update in the page :D

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat: add ManagedPackPage to the instance pages

Signed-off-by: flow <flowlnlnln@gmail.com>

aprilthepink commented:
Button to update Modpack automatically too?

aprilthepink commented:
PS: What about a notification that there are updates for a modpack?

flowln commented:
> Button to update Modpack automatically too?

i don't feel great about automatically updating a modpack, since people should really backup their stuff before an update. But the idea is to have a button that allows for easy updating if the user wants to :)

> PS: What about a notification that there are updates for a modpack?

maybe we could have something like that, but i guess you should make an issue about that, because it's unlikely i'll be able to do that in this single PR :p

aprilthepink commented:
> > Button to update Modpack automatically too?
> 
> i don't feel great about automatically updating a modpack, since people should really backup their stuff before an update. But the idea is to have a button that allows for easy updating if the user wants to :)
Yea, I did mean semi-automatic. You click the button and it does the update on its own. But you need to click it again for the next update.

> > PS: What about a notification that there are updates for a modpack?
> 
> maybe we could have something like that, but i guess you should make an issue about that, because it's unlikely i'll be able to do that in this single PR :p

Sure! I can even work on it as soon as you finished this PR myself :)

Scrumplex commented:
Is this scheduled for 5.0?

flowln commented:
> Is this scheduled for 5.0?

if we are going to release 5.0 today, no :iea:

flowln committed:
fix: use rich text in changelog for modrinth modpacks

Signed-off-by: flow <flowlnlnln@gmail.com>

ghost commented:
Some of my own opinions on this:
- make this menu a button in "version" instead, I'd rather not have any of my instances individually cursed/blessed with having a whole separate menu on the sidebar for modpack information.
- make sure an instance's modpack status can be completely removed by the user through the launcher

flowln commented:
> make this menu a button in "version" instead, I'd rather not have any of my instances individually cursed/blessed with having a whole separate menu on the sidebar for modpack information.

i don't think only a button is enough for this. It doesn't provide useful information such as changelogs and other metadata, which can in fact be very important when updating. My original design was to make it a tab inside a 'Overview' page, instead of being in the sidebar, but that'll take more time to be done. 

> make sure an instance's modpack status can be completely removed by the user through the launcher

Maybe, but currently this status doesn't add any restriction on its own, so if you don't want to use an instance as a managed pack, you can just not use the tab :p

Perhaps it could be an option in the future (similar to adding such a status to an already existing instance).

ghost commented:
> i don't think only a button is enough for this. It doesn't provide useful information such as changelogs and other metadata, which can in fact be very important when updating. My original design was to make it a tab inside a 'Overview' page, instead of being in the sidebar, but that'll take more time to be done.

Well, a button that opens the menu in a popup, that is.

flowln commented:
> Well, a button that opens the menu in a popup, that is.

why have it in a popup? it just seems unnecessary to me :/

ghost commented:
> 

To not clutter the sidebar and to not have wast amounts of emptiness in the window in case the edit instance menu is stretched out to cover the whole screen

flowln commented:
> To not clutter the sidebar and to not have wast amounts of emptiness in the window in case the edit instance menu is stretched out to cover the whole screen

![image](https://user-images.githubusercontent.com/9145768/197394380-0b439330-4d7f-4089-8c27-827c561f45e8.png)

i think adding something here would be more clutter than having a page on the sidebar though

ghost commented:
> 

we really should do something to this too though. we could swap the mod loader install section with the upper one, and move the folder shortcuts to the bottom.

Not to say there isn't clutter on the left.

ghost commented:
> Maybe, but currently this status doesn't add any restriction on its own, so if you don't want to use an instance as a managed pack, you can just not use the tab :p
> 
> Perhaps it could be an option in the future (similar to adding such a status to an already existing instance).

I wouldn't like to have 50% of my instances permanently marked as modpacks just because they are based off SO. There also needs to be an option to completely disable new modpack instances from getting modpack metadata.

The modpack updater should be treated in the same way the mod updater is - it should not be in your eyes too much, it's just something that you use once in a while and then forget about it for a few days. If it's not so, it's bloat.

ghost commented:
Is it for sure that the "if a new instance of a modpack is created, annoyingly prompt the suer to update an existing one" will be removed with this PR? Again, I want to update my modpack instance when i want to update it, not when i create a new instance of it.

flowln commented:
Dunno, some people seem to like it. It really depends on whether we can make the UX not be obnoxious for those that don't want it, but useful for those that do. Anyway, that'd not be done in this particular PR anyway, since it's not in the intended scope to change that.

ghost commented:
Can we make a toggle for this in the options? Like "Prompt to update modpack on instance creation"

flowln committed:
fix: do not display managed pack page for providers without an impl. yet

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(ManagedPackPage): better UX for when network requests fail / are pending

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(ManagedPackPage): add link to the Modrinth's pack page

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat+fix: allow forwarding extra info to InstanceImportTask

This allows us to pass to the creation instances their actual pack ID
and version ID, that in Flame's case, are only available before starting
to create an instance.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(ManagedPackPage): better changelog text when fetching / no changelog

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(ManagedPackPage): add Flame UI

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(ManagedPackPage): add warning about old bug with pack IDs

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(ManagedPackPage): give extra_info to InstanceImportTask in MR packs

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(ManagedPackPage): implement Flame modpack updating button

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(ManagedPackPage): clear combo boxes when adding new versions

Prevents versions to undergo mitosis.

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
chore(ManagedPackPage): format and add headers

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(ManagedPackPage): check app capabilities for creating the CF page

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
feat(ManagedPackPage): add reload button when in a fail state

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
refactor(ManagedPackPage): use smart pointers instead of raw ones

Signed-off-by: flow <flowlnlnln@gmail.com>

DioEgizio commented:
on curseforge packs, going to the curseforge tab, going back to version and then going back to curseforge causes a segmentation fault

flowln committed:
fix(ManagedPagePage): don't reload pack info when not needed

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln commented:
> on curseforge packs, going to the curseforge tab, going back to version and then going back to curseforge causes a segmentation fault

fixed, thanks!

RaptaG commented:
Lesgo it's not draft anymore!

Scrumplex commented on launcher/InstanceImportTask.cpp:
```diff
@@ -55,11 +55,9 @@
-    m_sourceUrl = sourceUrl;
-    m_parent = parent;
-}
+InstanceImportTask::InstanceImportTask(const QUrl sourceUrl, QWidget* parent, QMap<QString, QString> extra_info)
+    : m_sourceUrl(sourceUrl), m_extra_info(std::move(extra_info)), m_parent(parent)
+{}
```
What does `std::move` do here? Wouldn't we need to `std::move` while calling this constructor?
(`InstanceImportTask(coolUrl, coolParent, std::move(coolExtraInfo))`)

Scrumplex commented on launcher/ui/pages/instance/ManagedPackPage.cpp:
```diff
@@ -0,0 +1,432 @@
+    // We need to tell the user to redownload the pack, since we didn't save the required info previously
+    if (m_inst->getManagedPackID().isEmpty()) {
+        setFailState();
+        QString message =
+            tr("<h1>Hey there!</h1>"
+               "<h4>"
+               "It seems like your Pack ID is null. This is because of a bug in older versions of the launcher.<br/>"
+               "Unfortunately, we can't do the proper API requests without this information.<br/>"
+               "<br/>"
+               "So, in order for this feature to work, you will need to re-download the modpack from the built-in downloader.<br/>"
+               "<br/>"
+               "Don't worry though, it will ask you to update this instance instead, so you'll not lose this instance!"
+               "</h4>");
```
sad!

Scrumplex reviewed (approved):
I have seen the std::move pattern before as well. We could save another copy by maybe requiring rvalues for more constructors.

flowln committed:
refactor(InstanceImport): require rvalue from 'extra_info' mappings

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix(ManagedPackPage): preserve pack ID from Modrinth update

Signed-off-by: flow <flowlnlnln@gmail.com>

IThundxr reviewed (approved):
Looks good to me good work flow :)

Scrumplex reviewed (approved):
second approval :trollface:

stantios commented:
I got a problem. I updated from 4.2.1 to 4.4.4 then tried to update again to 4.5.4 but I keep getting this error: 
![image](https://user-images.githubusercontent.com/98621617/205374098-465e11c0-55bd-4e9e-a958-68565dd92c7a.png)

flowln commented:
@Imzxhir were you using the latest commit? i could swear i fixed this issue in [bdac799](https://github.com/PrismLauncher/PrismLauncher/pull/32/commits/bdac799500127c9d8ad25fa2c7d537c6c1186960) :pensive:

DioEgizio commented:
send prismlauncher-0.log

stantios commented:
> @Imzxhir were you using the latest commit? i could swear i fixed this issue in [bdac799](https://github.com/PrismLauncher/PrismLauncher/pull/32/commits/bdac799500127c9d8ad25fa2c7d537c6c1186960) 😔

I tried to but it didn't have a build with MacOS Legacy so I used the build before that one.

flowln commented:
I'm re-running the CI, hopefully it creates the artifacts correctly this time o.O

stantios commented:
> I'm re-running the CI, hopefully it creates the artifacts correctly this time o.O

Yep, it fixed the issue.

flowln committed:
fix(ManagedPackPage): only update the current instance exactly

Also carry on the original ID to avoid updating the wrong instance.

Signed-off-by: flow <flowlnlnln@gmail.com>

stantios commented:
I've got another bug where I can't copy instances, I'm not sure if it's related to this.
[PrismLauncher-1.log](https://github.com/PrismLauncher/PrismLauncher/files/10146529/PrismLauncher-1.log)

stantios commented:
I've got another bug to report as well where some mods from the modpack get removed. I downloaded the 4.2.0 version of Fabulously Optimized and updated to 4.4.4 and then when I check the mods, most of the mods got removed. 
[PrismLauncher-0.log](https://github.com/PrismLauncher/PrismLauncher/files/10146549/PrismLauncher-0.log)

flowln commented:
> I've got another bug where I can't copy instances, I'm not sure if it's related to this. [PrismLauncher-1.log](https://github.com/PrismLauncher/PrismLauncher/files/10146529/PrismLauncher-1.log)

I'm not sure, I didn't touch that code in this PR. Perhaps it's an issue with the library we use for `std::filesystem` on legacy macOS, because it seems to work fine on my end.

> I've got another bug to report as well where some mods from the modpack get removed. I downloaded the 4.2.0 version of Fabulously Optimized and updated to 4.4.4 and then when I check the mods, most of the mods got removed.
[PrismLauncher-0.log](https://github.com/PrismLauncher/PrismLauncher/files/10146549/PrismLauncher-0.log)

I did an oopsie, should hopefully be fixed in this new commit, thanks :)

stantios commented:
> > I've got another bug where I can't copy instances, I'm not sure if it's related to this. [PrismLauncher-1.log](https://github.com/PrismLauncher/PrismLauncher/files/10146529/PrismLauncher-1.log)
> 
> I'm not sure, I didn't touch that code in this PR. Perhaps it's an issue with the library we use for `std::filesystem` on legacy macOS, because it seems to work fine on my end.
> 
> > I've got another bug to report as well where some mods from the modpack get removed. I downloaded the 4.2.0 version of Fabulously Optimized and updated to 4.4.4 and then when I check the mods, most of the mods got removed.
> > [PrismLauncher-0.log](https://github.com/PrismLauncher/PrismLauncher/files/10146549/PrismLauncher-0.log)
> 
> I did an oopsie, should hopefully be fixed in this new commit, thanks :)

Ok, thanks. Also, maybe I could ask Dio about it since I think he wrote that part of code.

DioEgizio commented:
not related to that at all, i can't copy on macos (non-legacy) either

DioEgizio commented:
doesn't happen on 5.2, so it's a regression, but let's keep this out of here and in a separate issue

stantios commented:
> I've got another bug where I can't copy instances, I'm not sure if it's related to this. [PrismLauncher-1.log](https://github.com/PrismLauncher/PrismLauncher/files/10146529/PrismLauncher-1.log)

I'll create a separate issue for this

flowln committed:
fix(FlamePage): don't assert / suggest blocked modpacks

Signed-off-by: flow <flowlnlnln@gmail.com>

flowln committed:
fix: don't try updating Flame instance names when updating versions

Since the exact version string is only available in the manifest,
there's no easy way of getting it before commiting to the update, so
there's not much of a good way of showing the updated name in the UI,
and using the displayName is weird and gives some buggy behavior.

We may want to re-enable it in the future if we find a reliable way of
showing the correct info on the UI before starting the update.

Signed-off-by: flow <flowlnlnln@gmail.com>

Scrumplex reviewed (approved):
third approval :trollface:

ZekeZDev reviewed (approved):
Looks great to me, everything seems functional and I cant find any glaring issues. I do think that we will need to consider future changes to make updates more accessible however as the page is I think it does its job well.

RaptaG commented:
\ö/

lovemilk2333 commented:
Hello
if this function can upgrade custom-imported modpack?

DioEgizio commented:
> Hello
> if this function can upgrade custom-imported modpack?

holy necropost, please never do that again. you can already upgrade local modpacks if they're packaged as mrpack/cf

lovemilk2333 commented:
> > Hello  你好
> > if this function can upgrade custom-imported modpack?这个功能可以升级自定义导入的整合包吗？
> 
> holy necropost, please never do that again. you can already upgrade local modpacks if they're packaged as mrpack/cf我的天啊，这帖子都挖坟了，以后千万别再这么做了。如果本地整合包是 mrpack/cf 格式的，你现在就可以升级它们了。

Sorry
Could you provide a wiki link?

DioEgizio commented:
there's no wiki for it, but this is implemented for cf and modrinth

<img width="1231" height="838" alt="example" src="https://github.com/user-attachments/assets/278f1e85-1a17-49ce-a122-8cc30cd20486" />
